subprojects/gradle/build-trace/src/main/kotlin/com/avito/android/build_trace/BuildTracePlugin.kt

```diff
@@ -17,22 +16,36 @@ open class BuildTracePlugin : Plugin<Project> {
         check(project.isRoot()) {
             "Plugin must be applied to the root project but was applied to ${project.path}"
         }
-        if (isBuildTraceEnabled(project)) {
-            GradleCollector.initialize(project, listOf(buildTraceConsumer(project)))
+        val extension = project.extensions.create("buildTrace", BuildTraceExtension::class.java)
+
+        project.afterEvaluate {
+            registerListeners(project, extension)
         }
     }
 
-    private fun buildTraceConsumer(project: Project): BuildEventsListener = BuildTraceConsumer(
-        // TODO: pass it from an extension
-        output = File(project.projectDir, "outputs/trace/build.trace"),
-        loggerFactory = GradleLoggerFactory.fromPlugin(this, project)
-    )
-
-    // TODO: enable by a project extension
-    private fun isBuildTraceEnabled(project: Project): Boolean {
-        return project.buildEnvironment is BuildEnvironment.CI
-            || project.gradle.startParameter.isBuildScan
-            || project.gradle.startParameter.isProfile
-            || project.getBooleanProperty("android.enableProfileJson", default = false)
+    private fun registerListeners(project: Project, extension: BuildTraceExtension) {
+        val isEnabled = extension.enabled.getOrElse(false)
+        if (!isEnabled) return
+
+        val loggerFactory = GradleLoggerFactory.fromPlugin(this, project)
+
+        val criticalPathListener = criticalPathListener(project)
+
+        val buildTraceListener = BuildTraceListener(
+            output = File(pluginOutputDir(project), "build.trace"),
+            criticalPathProvider = criticalPathListener,
+            loggerFactory = loggerFactory
+        )
+        GradleCollector.initialize(project, listOf(criticalPathListener, buildTraceListener))
```

Comment:
Do we need to unregister listeners at the end of the build?

Reply:
Sure, this is the responsibility of `GradleCollector`.